estimagic/optimization/transform_problem.py

```diff
@@ -0,0 +1,529 @@
+import functools
+import json
+import warnings
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+from scipy.optimize._numdiff import approx_derivative
+
+from estimagic.decorators import expand_criterion_output
+from estimagic.decorators import handle_exceptions
+from estimagic.decorators import log_evaluation
+from estimagic.decorators import log_gradient
+from estimagic.decorators import log_gradient_status
+from estimagic.decorators import negative_criterion
+from estimagic.decorators import numpy_interface
+from estimagic.logging.create_database import prepare_database
+from estimagic.optimization.process_constraints import process_constraints
+from estimagic.optimization.reparametrize import reparametrize_to_internal
+from estimagic.optimization.utilities import index_element_to_string
+from estimagic.optimization.utilities import propose_algorithms
+
+
+def transform_problem(
+    criterion,
+    params,
+    algorithm,
+    criterion_kwargs,
+    constraints,
+    general_options,
+    algo_options,
+    gradient,
+    gradient_options,
+    logging,
+    log_options,
+    dashboard,
+    dash_options,
+):
+    """Transform the user supplied problem.
```

Comment:
This is a similar docstring as in _internal_minimize, which I changed. Please copy the changes. It makes sense to have duplication here, but we should keep the two versions aligned.

Reply:
Done in c56364a71dbbaf0122f4f136ffaedbeaa49b65cd.